File: app/src/main/java/com/github/malitsplus/shizurunotes/ui/charadetails/CharaDetailsFragment.java
package com.github.malitsplus.shizurunotes.ui.charadetails;

import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelProviders;

import android.app.Application;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.navigation.NavDirections;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.view.ContextMenu;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.github.malitsplus.shizurunotes.R;
import com.github.malitsplus.shizurunotes.databinding.FragmentCharaBinding;
import com.github.malitsplus.shizurunotes.databinding.FragmentCharaDetailsBinding;
import com.github.malitsplus.shizurunotes.ui.SharedViewModel;
import com.github.malitsplus.shizurunotes.ui.SharedViewModelFactory;
import com.github.malitsplus.shizurunotes.ui.charalist.CharaListFragmentDirections;

import java.lang.reflect.InvocationTargetException;

public class CharaDetailsFragment extends Fragment implements View.OnClickListener {

    private SharedViewModel sharedViewModel;
    private CharaDetailsViewModel detailsViewModel;
    private FragmentCharaDetailsBinding binding;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {

        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_chara_details, container, false);

        sharedViewModel = new ViewModelProvider(getActivity()).get(SharedViewModel.class);
        SharedViewModelFactory factory = new SharedViewModelFactory(sharedViewModel);
        detailsViewModel = new ViewModelProvider(this, factory).get(CharaDetailsViewModel.class);

        binding.setDetailsViewModel(detailsViewModel);
        binding.setClickListener(this);

        //set title
        ((AppCompatActivity)getActivity()).getSupportActionBar().setTitle(sharedViewModel.getSelectedChara().unitName);

        //角色技能顺序
        RecyclerView recyclerView = binding.attackPatternRecycler;
        GridLayoutManager layoutManager = new GridLayoutManager(getContext(), 6);
        AttackPatternAdapter adapter = new AttackPatternAdapter(getContext());
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.setAdapter(adapter);
        recyclerView.setNestedScrollingEnabled(false);

        //技能
        RecyclerView recyclerViewSkill = binding.skillRecycler;
        LinearLayoutManager layoutManagerSkill = new LinearLayoutManager(getContext());
        SkillAdapter adapterSkill = new SkillAdapter(getContext());
        recyclerViewSkill.setLayoutManager(layoutManagerSkill);
        recyclerViewSkill.setAdapter(adapterSkill);
        recyclerViewSkill.setNestedScrollingEnabled(false);

        //设置观察者
        detailsViewModel.getMutableChara().observe(getViewLifecycleOwner(), (chara) -> {
            binding.setDetailsViewModel(detailsViewModel);
            adapter.update(chara.attackPatternList.get(0).items);
            adapterSkill.update(chara.skills);

            if(chara.attackPatternList.size() == 1){
                binding.attackPatternConstraint.removeView(binding.textAnotherMode);
                binding.attackPatternConstraint.removeView(binding.attackPatternRecycler2);
            } else {
                RecyclerView recyclerView2 = binding.attackPatternRecycler2;
                GridLayoutManager layoutManager2 = new GridLayoutManager(getContext(), 6);
                AttackPatternAdapter adapter2 = new AttackPatternAdapter(getContext());
                recyclerView2.setLayoutManager(layoutManager2);
                recyclerView2.setAdapter(adapter2);
                recyclerView2.setNestedScrollingEnabled(false);
                adapter2.update(chara.attackPatternList.get(1).items);
            }
        });

        return binding.getRoot();
    }

    @Override
    public void onClick(View v){
        if(v.getId() == R.id.detailsItemChara){
            NavDirections action = CharaDetailsFragmentDirections.actionNavCharaDetailsToNavCharaProfile();
            Navigation.findNavController(v).navigate(action);
        }
    }

}

File: app/src/main/java/com/github/malitsplus/shizurunotes/ui/charadetails/CharaDetailsFragment.kt
package com.github.malitsplus.shizurunotes.ui.charadetails

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.Navigation
import androidx.navigation.findNavController
import androidx.navigation.fragment.navArgs
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.transition.TransitionInflater
import com.github.malitsplus.shizurunotes.R
import com.github.malitsplus.shizurunotes.data.Chara
import com.github.malitsplus.shizurunotes.databinding.FragmentCharaDetailsBinding
import com.github.malitsplus.shizurunotes.ui.shared.SharedViewModelChara
import com.github.malitsplus.shizurunotes.ui.shared.SharedViewModelCharaFactory
import com.github.malitsplus.shizurunotes.ui.base.AttackPatternContainerAdapter
import com.github.malitsplus.shizurunotes.ui.base.BaseHintAdapter
import com.github.malitsplus.shizurunotes.user.UserSettings

// TODO: 改成使用ViewType接口和适配器，避免NestedScrollView一次性渲染全部视图造成丢帧
class CharaDetailsFragment : Fragment(), View.OnClickListener {

    private lateinit var detailsViewModel: CharaDetailsViewModel
    private lateinit var sharedChara: SharedViewModelChara
    private lateinit var binding: FragmentCharaDetailsBinding
    private val args: CharaDetailsFragmentArgs by navArgs()

    private val adapterSkill by lazy { SkillAdapter(sharedChara) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        sharedChara = ViewModelProvider(requireActivity()).get(SharedViewModelChara::class.java)
        detailsViewModel = ViewModelProvider(this, SharedViewModelCharaFactory(sharedChara))[CharaDetailsViewModel::class.java]

        sharedElementEnterTransition =
            TransitionInflater.from(context)
                .inflateTransition(android.R.transition.move).setDuration(300)

        sharedElementReturnTransition =
            TransitionInflater.from(context)
                .inflateTransition(android.R.transition.move).setDuration(300)
    }

    override fun onResume() {
        super.onResume()
        binding.toolbar.menu.findItem(R.id.menu_chara_show_expression).isChecked = UserSettings.get().getExpression()
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {

        binding = FragmentCharaDetailsBinding.inflate(
            inflater,
            container,
            false
        ).apply {

            detailsItemChara.transitionName = "transItem_${args.charaId}"

            if (sharedChara.backFlag)
                appbar.setExpanded(false, false)

            detailsVM = detailsViewModel
            clickListener = this@CharaDetailsFragment
        }

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.apply {
            toolbar.setNavigationOnClickListener { view ->
                view.findNavController().navigateUp()
            }

            toolbar.setOnMenuItemClickListener {
                when (it.itemId) {
                    R.id.menu_chara_customize -> {
                        Navigation.findNavController(binding.root).navigate(
                            CharaDetailsFragmentDirections.actionNavCharaDetailsToNavAnalyze()
                        )
                    }
                    R.id.menu_chara_show_expression -> {
                        it.isChecked = !it.isChecked
                        UserSettings.get().setExpression(it.isChecked)
                        sharedChara.mSetSelectedChara(sharedChara.selectedChara)
                        adapterSkill.notifyDataSetChanged()
                    }
                }
                true
            }
        }

        // 技能循环
        val adapterAttackPattern = AttackPatternContainerAdapter(context).apply {
            initializeItems(detailsViewModel.mutableChara.value?.attackPatternList)
        }
        binding.attackPatternRecycler.apply {
            layoutManager = GridLayoutManager(context, 6).apply {
                spanSizeLookup = object : GridLayoutManager.SpanSizeLookup() {
                    override fun getSpanSize(position: Int): Int {
                        return when(adapterAttackPattern.getItemViewType(position)) {
                            BaseHintAdapter.HINT_TEXT -> 6
                            else -> 1
                        }
                    }
                }
            }
            adapter = adapterAttackPattern
        }

        // 技能 Recycler
        val layoutManagerSkill = LinearLayoutManager(context)
        binding.skillRecycler.apply {
            layoutManager = layoutManagerSkill
            adapter = adapterSkill
        }

        // 观察chara变化（1.0.0去掉rank下拉框后已经可以删掉了，留着备用）
        detailsViewModel.mutableChara.observe(
            viewLifecycleOwner,
            Observer<Chara> { chara: Chara ->
                binding.detailsVM = detailsViewModel
                adapterSkill.update(chara.skills)
            }
        )
    }

    override fun onClick(v: View?) {
        if (v?.id == R.id.detailsItemCharaContainer) {
            val action =
                CharaDetailsFragmentDirections.actionNavCharaDetailsToNavCharaProfile()
            Navigation.findNavController(v).navigate(action)
        }
    }
}